Fixed #19580 -- Made nullable relations work consistently

Rajesh-Veeranki opened the pull request:
Rebased the previous PR https://github.com/django/django/pull/7098/ with the master here. Modified the release text. The tests fail when I remove any newly added line. Please review and suggest any changes.

timgraham commented:
This doesn't address my comment from the other PR:  "Looking at the rest of it, I'm not sure what happened to much of Anssi's original patch (master...akaariai:ticket_17541). It seems that only the many-to-many changes survived the rebase?"

timgraham commented on docs/releases/1.11.2.txt:
```diff
@@ -39,3 +39,32 @@ Bugfixes
+
+.. _backwards-incompatible-1.11.2:
+
+Backwards incompatible changes in 1.11.2
```
This should target Django 2.0.

Rajesh-Veeranki commented:
You are right, the RelatedManager changes went missing. I added them back now. But more tests are required. I will add them in a few days.

Rajesh-Veeranki commented:
I have added changes and tests for RelatedManager. Please review them.

While writing the test cases, I came across this. When `Choice` is related to `Poll` by poll.question, i.e.


    class Poll(models.Model):
        question = models.CharField(max_length=200, unique=True, null=True)

        def __str__(self):
            return "Q: %s " % self.question


    class Choice(models.Model):
        poll = models.ForeignKey(Poll, models.SET_NULL, to_field='question', null=True)
        choice = models.CharField(max_length=200)

        def __str__(self):
            return "Choice: %s in poll %s" % (self.choice, self.poll)

Is this piece of code valid?
 
`p = Poll.objects.create(question=None)`
`c = Choice.objects.create(poll=p, choice='A')`
`p.choice_set.count() == 0`
The current django code evaluates them without error

timgraham commented on tests/m2m_through_regress/tests.py:
```diff
@@ -229,6 +230,15 @@ def test_m2m_relations_unusable_on_null_pk_obj(self):
+        nullcar = Car.objects.create(make=None)
+        with self.assertRaisesMessage(
+            ValueError,
+            '"<Car: None>" needs to have a value for field "make" '
```
Use the `msg  = ` style of the first test.

timgraham commented on tests/many_to_many/tests.py:
```diff
@@ -27,7 +27,7 @@ def setUp(self):
     def test_add(self):
         # Create an Article.
-        a5 = Article(headline='Django lets you reate Web apps easily')
+        a5 = Article(headline='Django lets you create Web apps easily')
```
Make this unrelated fix in a separate pull request.

timgraham commented on tests/many_to_one_null/models.py:
```diff
@@ -32,3 +32,18 @@ class Car(models.Model):
     car = models.ForeignKey(Car, models.SET_NULL, to_field='make', null=True, related_name='drivers')
+
+
+class Poll(models.Model):
```
Is there an issue reusing the `Car`/`Driver` (or some other) models? We prefer not to add new models unless absolutely necessary since that increases complexity and the time needed to run the tests.

carltongibson replied:
`Poll` and `Choice` can still be removed. The test using them just requires an unsaved model, so `Car()` triggers the error.

auvipy replied:
should we remove all the Poll and Choice models and related codes?

timgraham commented on tests/many_to_one_null/tests.py:
```diff
@@ -134,3 +134,25 @@ def test_related_null_to_field(self):
+        p = Poll.objects.create(question=None)
+        c = Choice.objects.create(choice='B')
+
+        with self.assertRaises(ValueError):
```
Use `assertRaiseMessage` as in the other tests.

timgraham commented on tests/null_queries/tests.py:
```diff
@@ -39,9 +39,9 @@ def test_none_as_null(self):
-        # Related managers use __exact=None implicitly if the object hasn't been saved.
         p2 = Poll(question="How?")
-        self.assertEqual(repr(p2.choice_set.all()), '<QuerySet []>')
+        with self.assertRaises(ValueError):
```
Use `assertRaiseMessage` as in the other tests (maybe this is better with the other added tests though).

timgraham commented on docs/releases/2.0.txt:
```diff
@@ -380,6 +380,28 @@ raises an exception and should be replaced with::
+key relations there was no error. To make the nullable relations consistent
+the following changes were made:
+
+* Nullable many-to-many and foreign key relations will return an empty
```
Some code examples to illustrate these points would be useful.

Rajesh-Veeranki replied:
Added them now. Please have a look.

timgraham commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -570,7 +576,15 @@ def get_prefetch_queryset(self, instances, queryset=None):
+        def _check_fk_val(self):
+            for field in self.field.foreign_related_fields:
+                if getattr(self.instance, field.attname) is None:
+                    raise ValueError('"%r" needs to have a value for field "%s" before '
```
Use same hanging indent style as line 516.

Rajesh-Veeranki commented on tests/many_to_one_null/models.py:
```diff
@@ -32,3 +32,18 @@ class Car(models.Model):
     car = models.ForeignKey(Car, models.SET_NULL, to_field='make', null=True, related_name='drivers')
+
+
+class Poll(models.Model):
```
My bad! I could reuse the models. Fixed it now.

Rajesh-Veeranki commented:
Hi @carltongibson 
Thanks for the review. Apologies for my delayed response.
Please find my reply to the corresponding comments below

1. I added `_check_has_fk_val` to `_add_items`, `_remove_items`, `clear` only because all the other functions gets called through them. So it's deliberate.
2. Adjusted now
3. I didn't get this comment. Can you please elaborate. Do you want me to move all the tests to null_queries/tests.py? 
4. Please give me the text for the docs. I can update my PR with it.

carltongibson commented:
HI @Rajesh-Veeranki. 

Please: 

* Clear up the failure to the docs build. 
* Move the relevant check from `null_queries/tests.py` to `many_to_one_null/tests.py` so they're all grouped, [as per Tim's comment](https://github.com/django/django/pull/8570#discussion_r121480284)
* Squash and rebase, making sure [the commit message uses the past tense](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines).

If you then uncheck "Patch need improvement" on the ticket, I will have another look, and adjust the wording on the docs. 

Thanks for your effort!

Rajesh-Veeranki commented:
Hi @carltongibson 
Thanks for the review. I'll work on your suggestions soon and reach out in case I've any doubt.

carltongibson commented:
Hi @Rajesh-Veeranki. Are you up for finishing this PR? (If not I can de-assign it and I'm sure someone would pick it up to get it across the line.)

LuRsT commented:
Gentle reminder @Rajesh-Veeranki, are you planning on picking this up again?

Rajesh-Veeranki commented:
Hi @LuRsT , I am NOT planning to pick this up anytime soon. So feel free to deassign. Thanks
